src/psygnal/_signal.py

```diff
@@ -996,7 +1031,32 @@ def __call__(
             check_types=check_types,
         )
 
-    def _run_emit_loop(self, args: tuple[Any, ...]) -> None:
+    # def _run_emit_loop(self, args: tuple[Any, ...]) -> None:
+    #     if self._recursion_mode == "immediate":
+    #         self._run_emit_loop_immediate(args)
+    #     else:
+    #         self._run_emit_loop_deferred(args)
+
+    def _run_emit_loop_immediate(self, args: tuple[Any, ...]) -> None:
+        # allow receiver to query sender with Signal.current_emitter()
+        with self._lock:
+            with Signal._emitting(self):
+                for caller in self._slots:
+                    try:
+                        caller.cb(args)
+                    except RecursionError as e:
+                        raise RecursionError(
+                            f"RecursionError in {caller.slot_repr()} when "
+                            f"emitting signal {self.name!r} with args {args}"
+                        ) from e
+                    except Exception as e:
+                        raise EmitLoopError(
+                            cb=caller, args=args, exc=e, signal=self
+                        ) from e
```

Comment:
I am not convinced that chain of few hundreds exceptions (3k recursion limit) is a good idea. Maybe information should be added once.

Reply:
i changed the logic a bit to wrap the inner bit (calling `caller.cb`) to unify the exception handling between the two modes.  I'm going to follow up on that in a follow up PR, but for this PR, I think they should now both be behaving like the loop you defined in #281